Introduce native ParquetReader

yingsu00 opened the pull request:
This commit introduces the native ParquetReader, which loads the file
metadata. It makes use of Apache Thrift included in Arrow, therefore
Arrow support was turned on when Parquet is enabled.

Depends on https://github.com/facebookincubator/velox/pull/2026

majetideepak commented:
>  It makes use of Apache Thrift included in Arrow.

Building Arrow just for thrift is overkill. Building thrift as an `ExternalProject` in Velox should be straightforward.

Yuhta commented on CMakeLists.txt:
```diff
@@ -121,6 +123,17 @@ endif()
+endif()
+
+if(VELOX_ENABLE_ARROW)
+  set(ARROW_PREFIX "${CMAKE_CURRENT_BINARY_DIR}/third_party/arrow_ep")
```
@yingsu00 My suggestion here is just wrapping the `thrift` that comes with Arrow in a nicer way.  No need to download it as a top level external project.  Although with my suggestion it would be easier later if we decide to bring `thrift` as a top level external project later.

yingsu00 replied:
@Yuhta Jimmy thank you. I wonder if we do as https://github.com/Yuhta/velox/blob/parquet-pr/third_party/CMakeLists.txt#L35-L39, can we also build the thrift-compiler then call it in dwio/parquet/thrift/CMakeLists.txt to generate the Parquet type .cpp/.h file? In third_party/arrow_ep/src/arrow_ep-build/thrift_ep-prefix/src/thrift_ep/CMakeLists.txt, I see it depends on `BUILD_COMPILER` but don't know how to pass it in. Any ideas? 
```
if(BUILD_COMPILER)
    if(NOT EXISTS ${THRIFT_COMPILER})
        set(THRIFT_COMPILER $<TARGET_FILE:thrift-compiler>)
    endif()
    add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/compiler/cpp)
elseif(EXISTS ${THRIFT_COMPILER})
    add_executable(thrift-compiler IMPORTED)
    set_property(TARGET thrift-compiler PROPERTY IMPORTED_LOCATION ${THRIFT_COMPILER})
endif()
```

Yuhta replied:
I would say we can checked in the thrift generated code for now, this will probably simplify the internal buck build a lot.

yingsu00 commented:
@Yuhta Jimmy, I tried the code block you suggested and it ran fine on my laptop. However CircleCI fails with the following:
```
CMake Error in velox/vector/arrow/tests/CMakeLists.txt:
  Imported target "arrow" includes non-existent path

    "/root/project/_build/debug/third_party/arrow_ep/src/arrow_ep-build/thrift_ep-install/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
make[1]: *** [Makefile:68: cmake] Error 1
make[1]: Leaving directory '/root/project'
make: *** [Makefile:81: debug] Error 2
```
The same error also happens on my laptop if I just run "make clean" in CLion and not delete the cmake-build-debug/third_party/arrow_eq folder.  The error went away after I removed the third_party/arrow_eq completely and then rebuild the velox_arrow_bridge_test. Do you have any idea why? Many thanks!

Yuhta commented:
@yingsu00 Not sure why, you probably want to check where the include directory is in CircleCI and adjust the path.

yingsu00 commented:
@Yuhta Jimmy I couldn't figure out why this PR fails with the change you suggested. The paths look good to me. I'm going to close this one in favor of https://github.com/facebookincubator/velox/pull/2054 if you are ok with it.

Yuhta commented on CMakeLists.txt:
```diff
@@ -14,6 +14,8 @@
 cmake_minimum_required(VERSION 3.10)
 set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/CMake" ${CMAKE_MODULE_PATH})
 
+add_definitions(-DNDEBUG)
```
This is no longer needed

Yuhta commented on velox/dwio/common/tests/utils/DataFiles.cpp:
```diff
@@ -26,12 +28,19 @@ bool endsWith(const std::string& s, const std::string& suffix) {
 
 std::string getDataFilePath(
-    const std::string& baseDir,
+    const std::string& pathFromFbcode,
```
Maybe we don't want to expose this name

yingsu00 replied:
@Yuhta Jimmy, thanks for reviewing the PR. Did you mean use another name that does not include "fb"? The code is like the following.
```
std::string getDataFilePath(
    const std::string& pathFromFbcode,
    const std::string& filePath) {
  std::string current_path = fs::current_path().c_str();
  if (endsWith(current_path, "fbcode")) {
    return current_path + "/" + pathFromFbcode + "/" + filePath;
  }
  return current_path + "/" + filePath;
}
```
If "fb" is a concern here, it did already include `endsWith(current_path, "fbcode")`. I actually don't know why this function was this way, but the baseDir was very confusing, since it is not used in my dev environement at all, and baseDir makes people think the path is baseDir + filePath but it's actually not. I created the function below that doesn't have this argument. Adn the renaming of pathFromFbcode is just to remove confusions. Do you have a recommendation what we should use here?

Yuhta replied:
I see, this is a hack someone has done to make it run with buck.  I was concerning about the name `fbcode`, but since it already appears in the code, it's ok to leave it like that.

majetideepak commented on velox/dwio/parquet/thrift/ParquetThriftTypes.h:
```diff
@@ -0,0 +1,3744 @@
+ */
+
+/**
+ * Autogenerated by Thrift Compiler (0.14.1)
```
Can we avoid checking in generated code?

Yuhta replied:
I think it's ok for now since this code does not tend to change, and it simplifies our build a lot.

yingsu00 commented:
@Yuhta Thank you for importing this PR! I just rebased it and removed the old commit of fixing the nullability warnings. The latest of that fix is in https://github.com/facebookincubator/velox/pull/2026 and we need to import that one first.

Yuhta commented:
@yingsu00 #2026 is small and we can do both at same time

yingsu00 commented:
@Yuhta Thank you very much Jimmy!
